core/src/main/java/org/apache/calcite/sql/SqlSignal.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.util.ImmutableNullableList;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Parse tree for {@code SqlSignal} call.
+ */
+public class SqlSignal extends SqlCall {
+  private static final SqlSpecialOperator OPERATOR =
+      new SqlSpecialOperator("SIGNAL", SqlKind.SIGNAL);
+
+  public final SignalType signalType;
+  public final SqlNode conditionOrSqlState;
+  public final SqlSetStmt setStmt;
+
+  /**
+   * Creates a {@code SqlSignal}.
+   *
+   * @param pos Parser position, must not be null
+   * @param signalType Specifies if call is SIGNAL or RESIGNAL, must not be null
+   * @param conditionOrSqlState A condition name or SQLSTATE value
+   * @param setStmt The optional SET clause
+   */
+  public SqlSignal(SqlParserPos pos, SignalType signalType,
+      SqlNode conditionOrSqlState, SqlSetStmt setStmt) {
+    super(pos);
+    this.signalType = Objects.requireNonNull(signalType);
+    this.conditionOrSqlState = conditionOrSqlState;
```

Comment:
nit: Use the UNSPECIFIED pattern and check that this is non null.

Reply:
Did you mean the signalType because the conditionOrSqlState is a SqlNode? There's already a non null check for signal type since those are the only two options

Reply:
Ah, you're right. I meant the signalType, which does already have the requireNonNull check.